src/services/public_http_server/handlers/relay_webhook/mod.rs

```diff
@@ -0,0 +1,199 @@
+        jwt::{JwtError, VerifyableClaims},
+        rpc::{
+            msg_id::get_message_id, Receipt, WatchAction, WatchEventClaims, WatchStatus, WatchType,
+            WatchWebhookPayload,
+        },
+    },
+    serde_json::json,
+    std::{collections::HashSet, sync::Arc, time::Instant},
+    thiserror::Error,
+    tracing::{error, info, instrument, warn},
+};
+
+pub mod error;
+pub mod handlers;
+
+#[derive(Debug, Error)]
+pub enum ClientError {
+    #[error("Could not parse watch event claims: {0}")]
+    ParseWatchEvent(JwtError),
+
+    #[error("Could not verify watch event: {0}")]
+    VerifyWatchEvent(JwtError),
+
+    #[error("JWT has wrong issuer")]
+    WrongIssuer,
+
+    #[error("Expected WatchAction::WatchEvent, got {0:?}")]
+    WrongWatchAction(WatchAction),
+
+    #[error("Expected WatchType::Subscriber, got {0:?}")]
+    WrongWatchType(WatchType),
+
+    #[error("Expected WatchStatus::Queued, got {0:?}")]
+    WrongWatchStatus(WatchStatus),
+}
+
+#[derive(Debug, Error)]
+pub enum Error {
+    #[error("Client error: {0}")]
+    ClientError(ClientError),
+
+    #[error("Server error: {0}")]
+    ServerError(RelayMessageServerError),
+}
+
+impl IntoResponse for Error {
+    fn into_response(self) -> Response {
+        match self {
+            Error::ClientError(e) => {
+                warn!("Relay webhook client error: {e:?}");
+                (
+                    StatusCode::UNPROCESSABLE_ENTITY,
+                    Json(json!({ "error": e.to_string() })),
+                )
+                    .into_response()
+            }
+            Error::ServerError(e) => {
+                error!("Relay webhook server error: {e:?}");
+                (
+                    StatusCode::INTERNAL_SERVER_ERROR,
+                    Json(json!({ "error": "Internal server error" })),
+                )
+                    .into_response()
+            }
+        }
+    }
+}
+
+pub async fn handler(
+    State(state): State<Arc<AppState>>,
+    Json(payload): Json<WatchWebhookPayload>,
+) -> Result<Response, Error> {
+    let claims = WatchEventClaims::try_from_str(&payload.event_auth)
+        .map_err(|e| Error::ClientError(ClientError::ParseWatchEvent(e)))?;
+
+    claims
+        .verify_basic(&HashSet::from([state.config.notify_url.to_string()]), None)
+        .map_err(|e| Error::ClientError(ClientError::VerifyWatchEvent(e)))?;
+
+    if claims.basic.iss != state.relay_identity {
+        return Err(Error::ClientError(ClientError::WrongIssuer));
+    }
+
+    // TODO check sub
+
+    let event = claims.evt;
+    let incoming_message = RelayIncomingMessage {
+        topic: event.topic,
+        message: event.message,
+        tag: event.tag,
+    };
+
+    // TODO send to channel and process in batches
+    state
+        .relay_client
+        .batch_receive(vec![Receipt {
+            topic: incoming_message.topic.clone(),
+            message_id: serde_json::from_str("0").unwrap(), // FIXME use actual message ID
```

Comment:
Should this be fixed as part of this PR? The relay is now sending these IDs in the payload.

Reply:
Yup, did in [1953773](https://github.com/WalletConnect/notify-server/pull/339/commits/1953773db51e44d2d27f9530629d86b1905bcf48)